### PR TITLE
add flux job drain

### PR DESCRIPTION
**Pull request by garlick**
Add a "drain" method to the job manager which 1) shuts off new job submissions, then 2) waits for the queue size to reach zero.  This supports the use case of an initial program that submits some jobs and exits as soon as they are all transitioned to INACTIVE.

Add a corresponding command `flux job drain [--timeout SECONDS]` which uses the drain RPC.  No API wrapper as yet - not sure if there will be a need?  (Easy to add if needed).

This was described in #2044.  I did make two changes from that description:
* `flux job drain --timeout N` was implemented with `flux_future_wait_for()`, not built in to the protocol.
* Multiple waiters are allowed.

### CODE REVIEW

**Comment by grondo** [comment]
Very cool!

Should there be a way to cancel a drain?

It does seem like one use case might be to submit a set of work, wait for all jobs to complete, then "undrain" and do it over again (or perhaps once the queue has been emptied after  a drain, it is open for business again by default)

I also wasn't sure what should happen if you ctrl-c out of a `flux job drain`, seems like it should cancel the drain if there are no more waiters, but I don't have a strong opinion on that.

**Comment by garlick** [comment]
I think maybe a `flux undrain` might be a good addition.

On ctrl-c or timeout, it feels to me like that's aborting the wait, but one would expect the queue disable to have already occurrred?  We can always change that behavior later once we get a feel for the tool, if it becomes evident that I am mistaken.

**Comment by grondo** [comment]
> On ctrl-c or timeout, it feels to me like that's aborting the wait, but one would expect the queue disable to have already occurrred?

Yeah, that makes sense. I think the only reason the behavior comes into question is because `flux job drain` is really two operations, one of which is canceled with ctrl-c and the other which is not. It is probably better if ctrl-c and timeout behave the same, so I think your current approach is correct (probably won't be called interactively anyway).

File: src/modules/job-manager/drain.c [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                                    bool *submit_disable)
+{
+    struct drain_ctx *ctx;
+
+    if (!(ctx = calloc (1, sizeof (*ctx))))
+        return NULL;
+    ctx->h = h;
+    ctx->queue = queue;
+    ctx->submit_disable = submit_disable;
+    if (!(ctx->requests = zlist_new ())) {
```
Comment (grondo) [inline src/modules/job-manager/drain.c]: I like how the various job-manager functions are split into their own source files, but in this case I just noticed that a lot of state from the main job-manager "context" structure has to be copied here, including a reference to the `submit_disable` bool.

Would less copying be required, and the use of the `bool *` be avoided if the job-manager ctx_t was shared with its sub-modules (via an internal header file)? You could make the context opaque and have operations like `struct queue * job_manager_get_queue (struct job_manager *jm)`.

Alternately, in this case the abstracted "drain" context could just be the request list if the msg handler lived in the main job-manager code.

This also might just be personal preference, in which case you can ignore the comment. I only noticed it because I was trying to figure out what `bool* submit_disable` was doing...

Comment (garlick) [reply src/modules/job-manager/drain.c]: I see your point.  Hmm.

I'm not feeling strongly one way or the other on how those "contexts" should be factored (pulled from the main or pushed from there).

I think maybe the strongest point you've raised is the bare `submit_disable` bool pointer, passed without explanation.  Perhaps it would be better to pass the `submit_ctx *` (opaque) in to drain.c and implement a `submit_enable()` / `submit_disable()` interface for it to use.

I'll go with that and let's see how it turns out.

**Comment by garlick** [comment]
Just forced pushed with a bit of refactoring in the submit code, to implement `submit_enable()` and `submit_disable()` instead of the boolean pointer passing, as discussed above.

Also added `flux job undrain`.

**Comment by grondo** [comment]
I'm getting a hang in `t2202-job-manager.t` here once every 10 runs or so (on fluke):
```
expecting success: 
	flux job drain 2>drain2.err &
	flux job undrain &&
	wait &&
	grep re-enabled drain2.err

```

I think I see a `lt-flux job drain` process still running, so I'm guessing it is the `wait` that is hanging. Could `flux job drain` in background be running after `flux job undrain` in rare cases?

**Comment by garlick** [comment]
Thanks @grondo.  I took your advice and moved that test to a small python script.  LMK if the solution looks OK and I'll squash.

**Comment by grondo** [comment]
>  I took your advice and moved that test to a small python script. LMK if the solution looks OK and I'll squash.

LGTM! Though I'm not the best judge of python code.

I wasn't able to get the test to hang, so the fix seems robust.

**Comment by grondo** [comment]
Looks like one code path that is missed in testing is invoking the `drain_complete_cb` via the `queue->empty_cb` -- all the test cases seem to call `drain` with an empty queue. (This is from viewing coverage results).

**Comment by garlick** [comment]
Hopefully this last addition addresses the missing coverage.  I had to fix another issue so that the queue in that job-manager sharness test could actually reach zero size - there was missing logic to handle cleanup of jobs that were in the queue to be scheduled (but in the test never would be, since sched isn't loaded).  Then it seemed necessary to use create another python script to be able to ensure the order of a drain request followed by a cancel request.

**Comment by grondo** [comment]
Once #2104 goes in, is this PR ready for a merge?

**Comment by garlick** [comment]
Yeah, I'll squash it down a bit first though.

**Comment by garlick** [comment]
I have this squashed down and rebased on top of #2014.  I'll hold off pushing until 2014 is merged.

**Comment by garlick** [comment]
This is ready to go in IMHO once travis turns green.

**Comment by grondo** [comment]
LGTM, merged.

**Comment by garlick** [comment]
Thanks!
